src/alias.c

```diff
@@ -0,0 +1,130 @@
+#include "alias.h"
+#include "util/dict.h"
+
+AliasTable *AliasTable_g = NULL;
+
+AliasTable *AliasTable_New(void) {
+  return dictCreate(&dictTypeHeapStrings, NULL);
+}
+
+void IndexAlias_InitGlobal(void) {
+  AliasTable_g = AliasTable_New();
+}
+
+void IndexAlias_DestroyGlobal(void) {
+  if (!AliasTable_g) {
+    return;
+  }
+  dictRelease(AliasTable_g);
+  AliasTable_g = NULL;
+}
+
+int AliasTable_Add(AliasTable *table, const char *alias, IndexSpec *spec, int options,
+                   QueryError *error) {
+  // look up and see if it exists:
+  dictEntry *e, *existing = NULL;
+  e = dictAddRaw(table, (void *)alias, &existing);
+  if (existing) {
+    QueryError_SetError(error, QUERY_EINDEXEXISTS, "Alias already exists");
+    return REDISMODULE_ERR;
+  }
+  assert(e->key != alias);
+  e->v.val = spec;
+  if (!(options & INDEXALIAS_NO_BACKREF)) {
+    char *duped = rm_strdup(alias);
+    spec->aliases = array_ensure_append(spec->aliases, &duped, 1, char *);
+  }
+  return REDISMODULE_OK;
+}
+
+int AliasTable_Del(AliasTable *table, const char *alias, IndexSpec *spec, int options,
+                   QueryError *error) {
+  char *toFree = NULL;
+  // ensure that the item exists in the list
+  if (!spec->aliases) {
+    QueryError_SetError(error, QUERY_ENOINDEX, "Alias does not belong to provided spec");
+    return REDISMODULE_ERR;
+  }
+
+  ssize_t idx = -1;
+  for (size_t ii = 0; ii < array_len(spec->aliases); ++ii) {
+    if (spec->aliases[ii] && !strcasecmp(spec->aliases[ii], alias)) {
+      idx = ii;
+      break;
+    }
+  }
+  if (idx == -1) {
+    QueryError_SetError(error, QUERY_ENOINDEX, "Alias does not belong to provided spec");
+    return REDISMODULE_ERR;
+  }
+
+  if (!(options & INDEXALIAS_NO_BACKREF)) {
+    toFree = spec->aliases[idx];
+    size_t oldLen = array_len(spec->aliases);
+    spec->aliases = array_del(spec->aliases, idx);
+  }
+
+  int rc = REDISMODULE_OK;
+  if (dictDelete(table, alias) == DICT_ERR) {
+    QueryError_SetError(error, QUERY_ENOINDEX, "Alias does not belong to index");
```

Comment:
This should be an assert

Reply:
this is user input.

Reply:
But if you reach here it means that the alias exists in the spec list so how can it not be in the alias table?